pkg/graveler/graveler.go

```diff
@@ -1068,39 +1068,234 @@ func (g *KVGraveler) List(ctx context.Context, repositoryID RepositoryID, ref Re
 }
 
 func (g *KVGraveler) Commit(ctx context.Context, repositoryID RepositoryID, branchID BranchID, params CommitParams) (CommitID, error) {
-	return g.db.Commit(ctx, repositoryID, branchID, params)
+	// TODO(issue 3569) - kv implementation optimistic lock commit
+	var preRunID string
+	var commit Commit
+	var storageNamespace StorageNamespace
+	res, err := g.db.branchLocker.MetadataUpdater(ctx, repositoryID, branchID, func() (interface{}, error) {
+		isProtected, err := g.protectedBranchesManager.IsBlocked(ctx, repositoryID, branchID, BranchProtectionBlockedAction_COMMIT)
+		if err != nil {
+			return nil, err
+		}
+		if isProtected {
+			return nil, ErrCommitToProtectedBranch
+		}
+		repo, err := g.GetRepository(ctx, repositoryID)
+		if err != nil {
+			return "", fmt.Errorf("get repository: %w", err)
+		}
+		storageNamespace = repo.StorageNamespace
+
+		branch, err := g.db.GetBranch(ctx, repositoryID, branchID)
+		if err != nil {
+			return "", fmt.Errorf("get branch: %w", err)
+		}
+
+		// fill commit information - use for pre-commit and after adding the commit information used by commit
+		commit = NewCommit()
+
+		if params.Date != nil {
+			commit.CreationDate = time.Unix(*params.Date, 0)
+		}
+		commit.Committer = params.Committer
+		commit.Message = params.Message
+		commit.Metadata = params.Metadata
+		if branch.CommitID != "" {
+			commit.Parents = CommitParents{branch.CommitID}
+		}
+
+		if params.SourceMetaRange != nil {
+			empty, err := g.stagingEmpty(ctx, branch)
+			if err != nil {
+				return nil, fmt.Errorf("checking empty branch: %w", err)
+			}
+			if !empty {
+				return nil, ErrCommitMetaRangeDirtyBranch
+			}
+		}
+
+		preRunID = g.hooks.NewRunID()
+		err = g.hooks.PreCommitHook(ctx, HookRecord{
+			RunID:            preRunID,
+			EventType:        EventTypePreCommit,
+			SourceRef:        branchID.Ref(),
+			RepositoryID:     repositoryID,
+			StorageNamespace: storageNamespace,
+			BranchID:         branchID,
+			Commit:           commit,
+		})
+		if err != nil {
+			return "", &HookAbortError{
+				EventType: EventTypePreCommit,
+				RunID:     preRunID,
+				Err:       err,
+			}
+		}
+
+		var branchMetaRangeID MetaRangeID
+		var parentGeneration int
+		if branch.CommitID != "" {
+			branchCommit, err := g.RefManager.GetCommit(ctx, repositoryID, branch.CommitID)
```

Comment:
This flow needs to be changed - both creation of commit and drop involve the sealedTokens as well. Please add a TODO so we'll address it later